include/revng/PTML/CommentPlacementHelper.h

```diff
@@ -0,0 +1,354 @@
+      uint64_t AdjustedRHS = Another.Numerator * this->Denominator;
+      return AdjustedLHS <=> AdjustedRHS;
+    }
+    bool operator==(const Score &Another) const {
+      return operator<=>(Another) == std::strong_ordering::equal;
+    }
+  };
+
+public:
+  struct CommentView {
+    uint64_t CommentIndex;
+    bool LocationMatchesExactly;
+    Score Score;
+    const CommentLocationType *ExpectedLocation;
+  };
+  using CommentList = std::vector<CommentView>;
+
+private:
+  // This assumes that `Trait::StatementType` is cheap (and, more importantly,
+  // safe) to copy
+  std::unordered_map<StatementType, CommentList> ResultMap;
+  CommentList HomelessComments;
+
+public:
+  CommentPlacementHelper() = default;
+  CommentPlacementHelper(const model::Function &Function, auto &&Graph) {
+    if (Function.Comments().empty()) {
+      // No comments in this function, nothing to map
+      return;
+    }
+
+    // Compute a dominator tree for the given graph to reduce duplicate
+    // comments: if a suitable statement is already dominated by another one,
+    // we can skip that emission.
+    using DominatorTreeType = llvm::DominatorTreeBase<InternalNodeType, false>;
+    DominatorTreeType DominatorTree;
+    DominatorTree.recalculate(Graph);
+
+    std::vector<std::pair<Score, StatementLocationType>> Scores;
+    Scores.resize(Function.Comments().size());
+
+    // First build a score map - select the best location for each comment
+    for (auto Node : llvm::depth_first(DominatorTree.getRootNode())) {
+      for (const auto &Statement : Trait::getStatements(Node->getBlock())) {
+        StatementLocationType Location = Trait::getAddresses(Statement);
+
+        for (auto &&[Index, Comment] : llvm::enumerate(Function.Comments())) {
+
+          // Use the Tversky Index as the scoring function.
+          // Note the parameters controlling the weights of the sets: the higher
+          // a value is, the bigger the relative negative impact of
+          // the corresponding set having _extra_ elements compared to the other
+          // one.
+          //
+          // We want this because a line having unaccounted for addresses is a
+          // way bigger red flag that a (potentially user edited) comment having
+          // them.
+          auto Score = tverskyIndex<1, 2>(Comment.Location(), Location);
+          if (Score.has_value())
+            if (shouldReplace(Scores[Index], { *Score, Location }))
+              Scores[Index] = { *Score, Location };
+        }
+      }
+    }
+
+    // Process homeless comments so they don't interfere with the others
+    for (auto &&[I, Comment] : llvm::enumerate(Function.Comments()))
+      if (Scores[I].first.Numerator == 0)
+        HomelessComments.emplace_back(I, false, Score{}, &Comment.Location());
+
+    TreeVisitor Visitor{ ResultMap, Function, Scores };
+    for (const auto &N : llvm::children<DTNode>(DominatorTree.getRootNode()))
+      Visitor(N);
```

Comment:
Can't you `Visitor(DominatorTree.getRootNode())`? 🤔 
`TreeVisitor` is a bit cumbersome, it be nice to inline it/make a lambda.

Reply:
Actually this was a bug! This extra loop was never necessary.